### PR TITLE
Implement long range correction for LJ potential

### LINKED ISSUE #1139: Add long-range tail correction to Lennard-Jones potential
## Description

In working on https://github.com/mosdef-hub/reproducibility_study, @ramanishsingh noted that the Trappe-UA forcefield was getting different density results in the MD and MC engines. In order to compare to the paper in which this forcefield was developed, we must use a long-range correction method. These methods are common in other engines and will help users of atomistic forcefields to represent the forces as they are intended by the forcefield developers.

## Proposed solution

Equation 5 from https://doi.org/10.1021/jp980939v is the correction used by LAMMPS

### PR DESCRIPTION
**Pull request by jennyfothergill**
## Description

Adds analytical tail correction to the LJ pair potential according to eq 5 from https://doi.org/10.1021/jp980939v

## Motivation and context

This change is needed for the correct implementation of certain forcefields, e.g., the TrappeUA forcefield.

Resolves #1139 

## How has this been tested?

(WIP) Comparing potential energy and instantaneous pressure difference of identical system with and without tail correction to that obtained in GROMACS, LAMMPS, GOMC, Cassandra, and MCCCS-MN.

Engines | Energy w/ tail correction (kJ/mol) | Energy w/o tail correction (kJ/mol) | Energy difference (kJ/mol) | Pressure w/ tail correction (kJ/mol/nm^3) | Pressure w/o tail correction (kJ/mol/nm^3) | Pressure difference (kJ/mol/nm^3)
-- | -- | -- | -- | -- | -- | --
HOOMD | 459209.3835 | 459337.6130 | -128.2295 | 30389.01054 | 30393.02325 | -4.01271145
GROMACS 2021.1 | 459060.5625 | 459317.1563 | -256.5938 |   |   | 0
GROMACS 2020.6 | 459188.9063 | 459317.1563 | -128.2500 |   |   |  
LAMMPS | 459215.4639 | 459343.6617 | -128.1978 | 30388.44712 | 30392.45966 | -4.01253809
GOMC | 459216.0800 | 459340.8000 | -124.7200 |   |   | 0
Cassandra | 459210.0760 | 459338.3060 | -128.2300 |   |   | 0
MCCCS-MN | 459206.6527 | 459334.8815 | -128.2287 | 30489.28295 | 30491.28953 | -2.006581786

## Change log

```
- Add  energy and pressure tail correction to LJ Pair Potential according to eq 5 from https://doi.org/10.1021/jp980939v.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/master/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.
- [x] Make sure mode is "none" when using tail correction.
- [x] Validate against all engines.
- [ ] Improve performance.

### CODE REVIEW

**Comment by joaander** [comment]
Please also create an issue to document the feature request and the motivation for it.

File: hoomd/md/EvaluatorPairLJ.h
Comment (joaander) [inline hoomd/md/EvaluatorPairLJ.h:230]: Complete your PR first and then someone else will have to worry about it :)

It would be helpful if you included the analytical expression for the integral in the comment to help that developer.

File: hoomd/md/PotentialPair.h
Comment (joaander) [inline hoomd/md/PotentialPair.h:963]:
```suggestion
        .def_property("tail_correction", &T::getTailCorrectionEnabled, &T::setTailCorrectionEnabled)
```

File: hoomd/md/PotentialPair.h
Comment (joaander) [inline hoomd/md/PotentialPair.h:186]: Return 1/3 of \delta W in elements 0, 3, and 5 (return 0 for other elements). \delta W = \delta P * D * V - where D is the dimensionality of the system (2 or 3) and V is the box volume (or area).

File: hoomd/md/PotentialPair.h
Comment (joaander) [inline hoomd/md/PotentialPair.h:192]:
```if (m_tail_correction_enabled)
    {
    // whole function body
}
return 0
```

File: hoomd/md/EvaluatorPairLJ.h
Comment (jennyfothergill) [reply hoomd/md/EvaluatorPairLJ.h]: I think I've done what you've asked for--I added the equations in LaTeX in the comment.

**Comment by jennyfothergill** [comment]
@tommy-waltmann 
Sorry, I marked this PR as ready for review because I am hoping for guidance on what to do next. Josh is more familiar with this addition, so I have tagged him for review.

File: hoomd/md/PotentialPair.h
Comment (tcmoore3) [reply hoomd/md/PotentialPair.h:256]: Added.

**Comment by tcmoore3** [comment]
@joaander I made the changes that you suggested back in October and is ready for another review. Thanks for the very clear direction you gave. This PR still needs:
1. Unit tests
1. Validation test(s)
2. Benchmarks

I can add some basic unit tests to make sure things behave as expected. I'll defer to @jennyfothergill for the validation tests since she has already run some of those.

**Comment by joaander** [comment]
> @joaander I made the changes that you suggested back in October and is ready for another review. Thanks for the very clear direction you gave. This PR still needs:
> 
> 1. Unit tests
> 2. Validation test(s)
> 3. Benchmarks
> 
> I can add some basic unit tests to make sure things behave as expected. I'll defer to @jennyfothergill for the validation tests since she has already run some of those.

Yes, please add unit tests. The validation tests needed to check these should not be added to HOOMD as they take far too long to run for CI. Jenny's work on the reproducibility study counts as a the required validation test.

File: hoomd/md/PotentialPairGPU.h
Comment (joaander) [inline hoomd/md/PotentialPairGPU.h:85]:
```suggestion
```
Base class methods will be called automatically. We only need to implement this method specifically here if we add a GPU optimized implementation.

File: hoomd/md/pair/pair.py
Comment (joaander) [inline hoomd/md/pair/pair.py:274]: Add documentation that defines the functional form of the tail correction in LaTeX math and cite the source of the tail correction.

File: hoomd/md/pair/pair.py
Comment (joaander) [inline hoomd/md/pair/pair.py:271]:
```suggestion
        tail_correction (bool): Whether to apply the isotropic integrated long range tail correction.
```

File: hoomd/md/pytest/test_potential.py
Comment (joaander) [inline hoomd/md/pytest/test_potential.py:1175]: Use an `assert_allclose` regression check on the values of the energy and pressure.

**Review by joaander — CHANGES_REQUESTED** [review]
Thanks! Some changes are needed before merging.

File: hoomd/md/pair/pair.py
Comment (tcmoore3) [reply hoomd/md/pair/pair.py:301]: See [this commit](https://github.com/glotzerlab/hoomd-blue/pull/1138/commits/e2fbdd4041d7d4f2d8359823e5688692762fc8f7)

File: hoomd/md/PotentialPairGPU.h
Comment (tcmoore3) [reply hoomd/md/PotentialPairGPU.h]: That's what I had originally thought, but not declaring that here led to compilation errors of the sort reported [here](https://stackoverflow.com/questions/9941987/there-are-no-arguments-that-depend-on-a-template-parameter). This error is resolved by using the `this` pointer; see [this commit](https://github.com/glotzerlab/hoomd-blue/pull/1138/commits/44914300670fcc690a4df6c8e9e4280c8cb161fa) for the updated code.

File: hoomd/md/PotentialPairGPU.h
Comment (tommy-waltmann) [inline hoomd/md/PotentialPairGPU.h:212]: Is computing this correction in a GPU kernel hard to do? It would improve performance if this could be a part of the GPU compute forces kernel.

**Review by tommy-waltmann — COMMENTED** [review]
This looks to be in good shape. I only have one suggestion and then this can be merged assuming Jenny has already done the work outside of hoomd to validate this code.

File: hoomd/md/PotentialPairGPU.h
Comment (joaander) [reply hoomd/md/PotentialPairGPU.h:212]: In NVT simulations, this PR as is poses minimal performance issues. The added code will only trigger on timesteps when energies/pressures are logged. However, in NPT simulations, the pressure correction will trigger on every timestep.

The expensive part of the computation is determining the number density of each particle type in the system. After that, the computation is a very inexpensive computation (per type pair, not per particle). The goal in this PR is to get a functioning code. @tommy-waltmann you are welcome to contribute a GPU kernel in a separate PR to count the number of particles of each type. The alternative would be to somehow detect when the number of particles of each type might change and only flag the expensive computation when needed.

**Comment by joaander** [comment]
> This looks to be in good shape. I only have one suggestion and then this can be merged assuming Jenny has already done the work outside of hoomd to validate this code.

This project https://github.com/mosdef-hub/reproducibility_study is comparing to LAMMPS, GROMACS, GOMC, Cassandra, and MCCCS. Jenny wrote the long range correction code and is using it in that study.

**Review by tommy-waltmann — APPROVED** [review]

**Comment by tcmoore3** [comment]
I moved the type counting loop to the constructor of the `PotentialPair` class, and that helps ease the performance hit (same test parameters as before):

N_particles | Tail correction | Run 1 | Run 2 | Run 3 | Run 4 | Run 5 | Mean TPS
-- | -- | -- | -- | -- | -- | -- | --
10,000 | False | 9,453 | 9,575 | 9,682 | 9,479 | 9,492 | 9,536
10,000 | True | 5,952 | 6,047 | 5,975 | 5,995 | 5,971 | 5,988
100,000 | False | 1,948 | 1,965 | 1,974 | 1,960 | 1,975 | 1,964
100,000 | True | 1,629 | 1,438 | 1,490 | 1,612 | 1,252 | 1,484

**Review by joaander — APPROVED** [review]
Thanks!

**Comment by tcmoore3** [comment]
After some digging with @joaander , we found the origin of the performance hit. It comes down to the fact that we weren't calling cudaMemAdvise on the `m_params` array, which led to page thrashing. We fixed that in [this commit](https://github.com/glotzerlab/hoomd-blue/pull/1138/commits/c95711858ef4713283cb7ef47858587378231a31). After the fix, the performance hit is minimal with the tail corrections enabled (as it should be given the form of the correction). New benchmarks included below.

N_particles | Tail correction | Run 1 | Run 2 | Run 3 | Run 4 | Run 5 | Mean TPS
-- | -- | -- | -- | -- | -- | -- | --
10,000 | False | 9,568 | 9,583 | 9,690 | 9,693 | 9,325 | 9,572
10,000 | True | 9,267 | 9,291 | 9,311 | 9,333 | 9,348 | 9,310
100,000 | False | 1,961 | 1,957 | 1,968 | 1,967 | 1,949 | 1,960
100,000 | True | 1,956 | 1,957 | 1,957 | 1,960 | 1,958 | 1,958

**Pull request merged**
